File: backend/services/portal_agent_assignment_service.py
"""
Service layer for portal agent assignments
"""
import logging
from database.portal_agent_assignment_db import (
    get_portal_agents,
    assign_agent_to_portal,
    remove_agent_from_portal,
    set_portal_agents
)

logger = logging.getLogger(__name__)


async def get_portal_agents_impl(portal_type: str, tenant_id: str) -> list[int]:
    """
    Get agents assigned to a portal
    
    Args:
        portal_type: Portal type (doctor, student, patient)
        tenant_id: Tenant ID
        
    Returns:
        List of agent IDs
    """
    try:
        return get_portal_agents(portal_type, tenant_id)
    except Exception as e:
        logger.error(f"Failed to get portal agents: {str(e)}")
        raise ValueError(f"Failed to get portal agents: {str(e)}")


async def assign_agent_to_portal_impl(
    portal_type: str,
    agent_id: int,
    tenant_id: str,
    user_id: str
) -> dict:
    """
    Assign an agent to a portal
    
    Args:
        portal_type: Portal type (doctor, student, patient)
        agent_id: Agent ID
        tenant_id: Tenant ID
        user_id: User ID
        
    Returns:
        Assignment record
    """
    try:
        return assign_agent_to_portal(portal_type, agent_id, tenant_id, user_id)
    except Exception as e:
        logger.error(f"Failed to assign agent to portal: {str(e)}")
        raise ValueError(f"Failed to assign agent to portal: {str(e)}")


async def remove_agent_from_portal_impl(
    portal_type: str,
    agent_id: int,
    tenant_id: str,
    user_id: str
) -> None:
    """
    Remove an agent from a portal
    
    Args:
        portal_type: Portal type (doctor, student, patient)
        agent_id: Agent ID
        tenant_id: Tenant ID
        user_id: User ID
    """
    try:
        remove_agent_from_portal(portal_type, agent_id, tenant_id, user_id)
    except Exception as e:
        logger.error(f"Failed to remove agent from portal: {str(e)}")
        raise ValueError(f"Failed to remove agent from portal: {str(e)}")


async def set_portal_agents_impl(
    portal_type: str,
    agent_ids: list[int],
    tenant_id: str,
    user_id: str
) -> list[dict]:
    """
    Set complete list of agents for a portal
    
    Args:
        portal_type: Portal type (doctor, student, patient)
        agent_ids: List of agent IDs
        tenant_id: Tenant ID
        user_id: User ID
        
    Returns:
        List of assignment records
    """
    try:
        return set_portal_agents(portal_type, agent_ids, tenant_id, user_id)
    except Exception as e:
        logger.error(f"Failed to set portal agents: {str(e)}")
        raise ValueError(f"Failed to set portal agents: {str(e)}")

File: frontend/services/portalAgentAssignmentService.ts
import { API_ENDPOINTS } from "./api";
import { getAuthHeaders } from "@/lib/auth";
import log from "@/lib/logger";

/**
 * Portal Agent Assignment Service
 * Handles API calls for assigning agents to portals (doctor, student, patient)
 */

export type PortalType = "doctor" | "student" | "patient";

/**
 * Get list of agent IDs assigned to a portal
 * @param portalType Portal type (doctor, student, patient)
 * @returns List of agent IDs
 */
export const getPortalAgents = async (portalType: PortalType): Promise<number[]> => {
  try {
    const response = await fetch(API_ENDPOINTS.portalAgentAssignment.getAgents(portalType), {
      method: "GET",
      headers: getAuthHeaders(),
    });

    if (!response.ok) {
      throw new Error(`Failed to get portal agents: ${response.status}`);
    }

    const data = await response.json();
    return data.agent_ids || [];
  } catch (error) {
    log.error("Error getting portal agents:", error);
    throw error;
  }
};

/**
 * Assign an agent to a portal
 * @param portalType Portal type (doctor, student, patient)
 * @param agentId Agent ID to assign
 */
export const assignAgentToPortal = async (
  portalType: PortalType,
  agentId: number
): Promise<void> => {
  try {
    const response = await fetch(API_ENDPOINTS.portalAgentAssignment.assign, {
      method: "POST",
      headers: getAuthHeaders(),
      body: JSON.stringify({
        portal_type: portalType,
        agent_id: agentId,
      }),
    });

    if (!response.ok) {
      throw new Error(`Failed to assign agent: ${response.status}`);
    }
  } catch (error) {
    log.error("Error assigning agent to portal:", error);
    throw error;
  }
};

/**
 * Remove an agent from a portal
 * @param portalType Portal type (doctor, student, patient)
 * @param agentId Agent ID to remove
 */
export const removeAgentFromPortal = async (
  portalType: PortalType,
  agentId: number
): Promise<void> => {
  try {
    const response = await fetch(API_ENDPOINTS.portalAgentAssignment.remove, {
      method: "POST",
      headers: getAuthHeaders(),
      body: JSON.stringify({
        portal_type: portalType,
        agent_id: agentId,
      }),
    });

    if (!response.ok) {
      throw new Error(`Failed to remove agent: ${response.status}`);
    }
  } catch (error) {
    log.error("Error removing agent from portal:", error);
    throw error;
  }
};

/**
 * Set complete list of agents for a portal (replaces existing)
 * @param portalType Portal type (doctor, student, patient)
 * @param agentIds List of agent IDs
 */
export const setPortalAgents = async (
  portalType: PortalType,
  agentIds: number[]
): Promise<void> => {
  try {
    const response = await fetch(API_ENDPOINTS.portalAgentAssignment.setAgents, {
      method: "POST",
      headers: getAuthHeaders(),
      body: JSON.stringify({
        portal_type: portalType,
        agent_ids: agentIds,
      }),
    });

    if (!response.ok) {
      throw new Error(`Failed to set portal agents: ${response.status}`);
    }
  } catch (error) {
    log.error("Error setting portal agents:", error);
    throw error;
  }
};

